rust/agama-lib/src/scripts/model.rs

```diff
@@ -0,0 +1,215 @@
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 2 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE LLC about this file by physical or electronic mail, you may
+// find current contact information at www.suse.com.
+
+use std::{
+    fs,
+    io::Write,
+    os::unix::fs::OpenOptionsExt,
+    path::{Path, PathBuf},
+    process,
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::transfer::Transfer;
+
+use super::ScriptError;
+
+#[derive(Debug, Clone, Copy, PartialEq, strum::Display, Serialize, Deserialize)]
+#[strum(serialize_all = "camelCase")]
+#[serde(rename_all = "camelCase")]
+pub enum ScriptsGroup {
+    Pre,
+    Post,
+}
+
+/// Represents a script to run as part of the installation process.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Script {
+    /// Script's name.
+    pub name: String,
+    /// Script's body. Either the body or the URL must be specified.
+    pub body: Option<String>,
+    /// URL to get the script from. Either the body or the URL must be specified.
+    pub url: Option<String>,
+    /// Script's group
+    pub group: ScriptsGroup,
+}
+
+impl Script {
+    /// Runs the script and returns the output.
+    ///
+    /// * `workdir`: where to write assets (script, logs and exit code).
+    pub async fn run(&self, workdir: &Path) -> Result<(), ScriptError> {
+        let dir = workdir.join(self.group.to_string());
+
+        let path = dir.join(&self.name);
+        self.write(&path).await?;
+
+        let output = process::Command::new(&path).output()?;
+
+        let stdout_log = dir.join(format!("{}.log", &self.name));
+        fs::write(stdout_log, output.stdout)?;
+
+        let stderr_log = dir.join(format!("{}.err", &self.name));
+        fs::write(stderr_log, output.stderr)?;
+
+        let status_file = dir.join(format!("{}.out", &self.name));
+        fs::write(status_file, output.status.to_string())?;
+
+        Ok(())
+    }
+
+    /// Writes the script to the file system.
+    ///
+    /// * `path`: path to write the script to.
+    async fn write<P: AsRef<Path>>(&self, path: P) -> Result<(), ScriptError> {
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .mode(0o500)
+            .open(&path)?;
+
+        if let Some(url) = &self.url {
+            Transfer::get(url, file)?;
+        } else if let Some(body) = &self.body {
+            write!(file, "{}", &body)?;
+        }
+        // FIXME: else: invalid script definition
```

Comment:
as said above, should be already catched by json serializer and not here in code where you can rely on valid code.